Build: Migrate unit tests from Jest to Vitest

yannbf opened the pull request:
<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->


## What I did

This migrates the monorepo unit tests from Jest to Vitest.  This will make it easier to convert our codebase to ESM, and vitest has some good DX which we can take advantage of.

I wrote a codemod to do most of the manual labor and you can find it here: https://go.intuita.io/IX3EFT

<details>
<summary>The codemod takes care of:</summary>

1. JSDOM environment comments
```ts
// from
/**
 * @jest-environment jsdom
 */

// to
/**
 * @vitest-environment jsdom
 */
```

2. Type reference comments
```ts
// from
/// <reference types="@types/jest" />;

// to
/// <reference types="vitest" />;
```

3. All sorts of spy/mock/fn migrations
```ts
// from
jest.mock('@storybook/preview-api');
const value = jest.requireActual('@storybook/preview-api');
const value2 = jest.createMockFromModule('@storybook/preview-api');
const myMock = jest.fn();
const spy = jest.spyOn(EventsPackageExport, 'bla');

// to
import { vi } from "vitest";

vi.mock('@storybook/preview-api');
const value = await vi.importActual('@storybook/preview-api');
const value2 = vi.mock('@storybook/preview-api')
const myMock = vi.fn();
const spy = vi.spyOn(EventsPackageExport, 'bla');
```

4. All sorts of TypeScript type migrations
```ts
// from
const mockedApi = api as unknown as jest.Mocked<api.API>;
const toIdMock = toId as jest.Mock<ReturnType<typeof toId>>;
let mockChannel: { on: jest.Mock; emit?: jest.Mock };
(helpers.isNxProject as jest.SpyInstance).mockResolvedValue(true);

// to
import { Mock, Mocked, SpyInstance } from "vitest";

const mockedApi = api as unknown as Mocked<api.API>;
const toIdMock = toId as Mock<ReturnType<typeof toId>>;
let mockChannel: { on: Mock; emit?: Mock };
(helpers.isNxProject as SpyInstance).mockResolvedValue(true);
```

Apart from adding missing imports for `describe, it, test, beforeEach, beforeAll, afterEach, afterAll`, etc.

</details>

To sum the changes up:
- Codemod ran across all files in the codebase
- storyshots and storyshots-puppeteer packages are deleted. (they are deprecated, and in Storybook 8 they would be removed. Else we'd need to do a rework on them for proper ESM support)
- Mocking mechanism changed e.g.
- A lot of snapshots are updated
- Some package.json exports were changed
- jest config files were turned into vitest config files
- `jest-os-detection` (for windowsOnly tests) usage was replaced with a small local utility
- `jest-specific-snapshot` was replaced with Vitest's own snapshot utility

## Solved problems

We encountered and overcame some challenges in this conversion, see collapsed details below if you're interested.

<details>
<summary>details</summary>

### Out of memory issue - Solved!

~When running the tests, the process hangs for about 3min and then presents the following error message:
<img width="651" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/f85a3526-00e2-48d2-83c8-0287270ff62d">~

~This has to be investigated and fixed.~

Fixed by https://github.com/storybookjs/storybook/pull/24415/commits/6d35099f00af18056ec2dc01537b237f4acff3da

### Emotion loading twice - Solved!

It seems that @emotion/react is being loaded in two different contexts, which causes the theme to be lost, and some tests to fail. It's unclear what is causing this.  This is breaking tests in `addons/interactions`, `addons/a11y`, and `ui/blocks`, and `ui/manager`.

### Call counts in StoryIndexGenerator tests are off by one - Solved!

`AssertionError: expected "N" to be called 6 times, but got 5 times`.  It's unclear what caused this, or whether it is truly broken.  This breaks tests in `lib/core-server`.

</details>


### Skipped tests

We couldn't figure out an emotion theming issue for a couple of tests, so chose to temporarily skip them to unblock the rest of the conversion.

- `addons/a11y/src/components/A11YPanel.test.tsx`
- `addons/a11y/src/components/VisionSimulator.test.tsx`

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [x] integration tests
- [ ] end-to-end tests

#### Manual testing

Run `yarn test` or `yarn test:watch` in the `code` directory.

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

ndelangen committed:
add vitest to the monorepo - WIP

ndelangen committed:
remove storyshots some more and delete jest configs

ndelangen committed:
migrate first 2 stories by hand

ndelangen committed:
run tests only once

ndelangen committed:
migrate some things by hand

ndelangen committed:
more manual migrations

ndelangen committed:
making progress on migrating to vitest

yannbf committed:
migrate all jest references to vitest

yannbf committed:
update snapshots

yannbf committed:
more fixes

ndelangen committed:
progress on fixes

ndelangen committed:
Merge branch 'norbert/vitest-for-monorepo' of github.com:storybookjs/storybook into norbert/vitest-for-monorepo

ndelangen committed:
fixing tests in scripts

yannbf committed:
remove jest-specific-snapshot and update snapshots

yannbf committed:
fix imports

yannbf committed:
more test fixes

yannbf committed:
fix more tests

yannbf committed:
more fixes

ndelangen committed:
fix codemods test

ndelangen committed:
improve test

ndelangen committed:
fix a test

yannbf committed:
remove jest-os-detection

yannbf committed:
Merge branch 'next' into norbert/vitest-for-monorepo

yannbf committed:
upgrade to vitest 1.0.0-beta.4 and fix many more tests

yannbf committed:
Merge branch 'next' into norbert/vitest-for-monorepo

IanVS committed:
Add/configure jest-dom

IanVS committed:
Cleanup after rendering

We could create a render helper which does this automatically, perhaps,
but this seems clean enough, and there's only a handful of usages.

IanVS committed:
Only use vite 4

Previously vitest was also bringing in vite 5 beta,
which screwed up the types

IanVS committed:
Get renderers/vue tests passing

IanVS committed:
Get renderers/vue3 tests passing

IanVS committed:
Get renderers/svelte tests passing

IanVS commented:
Status update: I've gotten the `renderers` passing, and most of `addons`, except for a problem whereby emotion is somehow loaded twice, so the test loses the `theme` context.

IanVS commented on code/renderers/vue/package.json:
```diff
@@ -77,6 +79,9 @@
     "node": ">=16.0.0"
   },
+  "installConfig": {
+    "hoistingLimits": "workspaces"
```
I needed to add this to keep `vue-template-compiler` from being hoisted up to the monorepo root, where it would find vue3 (also hoisted) and crash.

JReinhold replied:
We're removing Vue2 in SB 8 anyway, so this doesn't matter.

IanVS committed:
Fix cli helpers test

IanVS committed:
Recreate throwMessage helpers from jest into vitest setup

IanVS committed:
Fix type error

IanVS committed:
Avoid empty mockImplementation

IanVS committed:
A bit cleaner way to deal with fsExtra.__setMockFiles

IanVS committed:
Avoid console warning error

IanVS committed:
Fix lib/codemod test

IanVS committed:
Use `import()` for presets instead of `interopRequireDefault`

vitest cannot mock `require` calls, but it can handle dynamic
imports quite nicely.  Since we require node 16+, I _think_ this
is safe, but it will need to be confirmed.

IanVS committed:
Get lib/core-common tests passing

IanVS committed:
Get lib/theming tests passing

If we're only using the @storyboook/globals package to help with
testing, we might be able to remove it now, as vitest can
handle global mocks quite well.

IanVS commented on code/lib/core-common/src/presets.ts:
```diff
@@ -210,8 +209,7 @@ async function getContent(input: any) {
   const name = input.name ? input.name : input;
-
-  return interopRequireDefault(name);
+  return import(name);
```
@ndelangen this is a bit of a risky change, and needs your 👀.

I do know that without this, it's going to be very difficult to use vitest, since it only mocks `import` and not `require`

JReinhold replied:
I've managed to revert this in https://github.com/storybookjs/storybook/pull/24415/commits/6dfcad5214340af3080d958777599aaa5efed612 and use `mock-require` to mock the presets instead. Seems to work

IanVS commented on code/addons/interactions/package.json:
```diff
@@ -43,6 +43,11 @@
+    "./preset": {
+      "types": "./dist/preset.d.ts",
+      "require": "./dist/preset.js",
+      "import": "./dist/preset.js"
```
This `import` condition and the one above it are not correct, they're pointing to cjs files.  But we don't generate an ESM version, so I did this just to get tests passing.  It should probably be fixed the right way by generating esm.

IanVS commented on code/lib/core-common/src/presets.test.ts:
```diff
@@ -26,7 +41,7 @@ vi.mock('./utils/safeResolve', () => {
     '@storybook/react',
     '@storybook/addon-actions/manager',
-    '@storybook/addon-actions/register',
+    '@storybook/addon-actions/register.js',
```
We don't have a condition for `/register`, only `./register.js`

IanVS commented on code/lib/theming/src/utils.ts:
```diff
@@ -63,9 +60,9 @@ export const darkenColor = colorFactory('darken');
 // scheme is set to dark we always want to use the light theme
 export const getPreferredColorScheme = () => {
-  if (!globalWindow || !globalWindow.matchMedia) return 'light';
+  if (!globalThis || !globalThis.matchMedia) return 'light';
```
I believe `globalThis` is supported in all node and browser versions that we are currently targeting, and it makes mocking in the tests much easier.

JReinhold replied:
It would be interesting to try and replace `@storybook/global` with `globalThis` everywhere someday.

IanVS commented on code/lib/theming/src/tests/util.test.js:
```diff
@@ -100,9 +100,12 @@ describe('utils', () => {
+        vi.fn().mockReturnValue({
+          matches: true,
+        })
+      );
```
We can also just use `globalThis.matchMedia = ` directly, but then we can't use `vi.unstubAllGlobals` if we need it.  (ref https://vitest.dev/api/vi.html#vi-stubglobal)

IanVS committed:
Update snapshot format

IanVS committed:
Avoid infinite loop in find-up

When we muck with process.cwd(), find-up can easily go into an infinite
loop.  Since we don't care about finding .git folders and such in this test, we can
just mock out the `getProjectRoot` to return the (fake) cwd.

IanVS committed:
Be consistent about named vs default imports

IanVS committed:
Use default actual import from debounce

IanVS committed:
Use vi.waitFor() instead of timing hacks

IanVS committed:
[HACK] add import condition for @storybook/preview/globals

IanVS committed:
Use arg of mock factory to clean up types

IanVS committed:
Use vi.mocked() instead of casting to Mock

IanVS committed:
Remove some unused code

IanVS commented:
I've done what I can for now, I need to step back and work on other things, but hopefully someone else can pick up the mantle and get this across the finish line.

JReinhold committed:
fix StoryIndexGenerator.test

JReinhold committed:
fix StoryIndexGenerator.deprecated.test

JReinhold commented:
I've investigated and fixed the tests related to `StoryIndexGenerator`.
The tests mocked a function in a separate package, and counted how many times that mock was called.

In Jest the mock was applied both in the module under test, but also any sub-packages that were being invoked had the mock as well. In Vitest however, sub-packages that use the same mocked package for some reason don't get the mock function, they still call the real function.

This caused the mock call count to be off by one between Jest and Vitest. I changed the expected count, because the count is not actually what's important in the tests, so the tests still asserts what's needed.

ndelangen committed:
add ignore

ndelangen committed:
Merge branch 'release-8-0' into norbert/vitest-for-monorepo

ndelangen committed:
ts fixes

ndelangen committed:
upgrade @testing-library/jest-dom

ndelangen committed:
update snapshots && fixes

ndelangen committed:
fixes

ndelangen committed:
upgrade vitest to beta.3 & fixes

ndelangen commented:
@SocketSecurity ignore @vitest/snapshot@0.34.6
@SocketSecurity ignore @vitest/snapshot@1.0.0-beta.3
@SocketSecurity ignore vue-template-compiler@2.6.14
@SocketSecurity ignore tinypool@0.8.1
@SocketSecurity ignore vitest@0.34.6
@SocketSecurity ignore vitest@1.0.0-beta.3

ndelangen committed:
fix linting

ndelangen committed:
fix check

ndelangen committed:
fix check of scripts

ndelangen committed:
remove invalid flags

ndelangen committed:
remove --ci flags

ndelangen committed:
add @vitest/coverage-v8 - npm

ndelangen commented:
I tried... 😢

ndelangen committed:
add package

IanVS commented:
@ndelangen what is left to fix here?

JReinhold committed:
remove --runInBand test flag

JReinhold commented:
@IanVS I've updated the list of failing tests. It looks like basing this on `release-8-0` instead of `next` has caused more tests to fail with the double-emotion loading error, most likely because of the new layout architecture.

JReinhold committed:
set vi globals to silent emotion warnings

JReinhold commented:
I've fixed the emotion double loading error. This hint was here:

https://github.com/emotion-js/emotion/pull/2905

This is a known issue in emotion, and they solve it by checking for the `vi` global before showing the warning. However [Vitest doesn't include that global by default](https://vitest.dev/config/#globals), but including it fixed the issue. (I don't think we should use the global, so I haven't configured TS to know about it)

IanVS committed:
Merge remote-tracking branch 'origin/next' into norbert/vitest-for-monorepo

# Conflicts:
#	MIGRATION.md
#	code/addons/a11y/package.json
#	code/addons/actions/package.json
#	code/addons/backgrounds/package.json
#	code/addons/interactions/package.json
#	code/addons/jest/package.json
#	code/addons/measure/package.json
#	code/addons/outline/package.json
#	code/addons/storyshots-core/package.json
#	code/addons/storyshots-puppeteer/package.json
#	code/addons/storysource/package.json
#	code/addons/themes/package.json
#	code/addons/toolbars/package.json
#	code/addons/viewport/package.json
#	code/builders/builder-vite/src/optimizeDeps.ts
#	code/deprecated/addons/package.json
#	code/deprecated/channel-postmessage/package.json
#	code/deprecated/channel-websocket/package.json
#	code/deprecated/client-api/package.json
#	code/deprecated/core-client/package.json
#	code/deprecated/manager-api-shim/package.json
#	code/deprecated/preview-web/package.json
#	code/deprecated/store/package.json
#	code/frameworks/nextjs/src/routing/app-router-provider.tsx
#	code/frameworks/nextjs/src/routing/page-router-provider.tsx
#	code/jest.config.js
#	code/lib/cli/package.json
#	code/lib/cli/scripts/generate-sb-packages-versions.js
#	code/lib/cli/src/versions.ts
#	code/lib/docs-tools/package.json
#	code/lib/manager-api/package.json
#	code/lib/preview-api/package.json
#	code/lib/preview/package.json
#	code/lib/preview/src/globals/globals.ts
#	code/lib/preview/src/globals/runtime.ts
#	code/package.json
#	code/ui/manager/src/globals/runtime.ts
#	code/ui/manager/src/globals/types.ts
#	code/ui/manager/src/runtime.ts
#	code/ui/manager/src/typings.d.ts
#	code/yarn.lock
#	docs/essentials/actions.md
#	docs/essentials/interactions.md
#	docs/writing-tests/snapshot-testing.md
#	scripts/package.json
#	scripts/prepare/bundle.ts
#	scripts/prepare/check-scripts.ts
#	scripts/prepare/check.ts
#	scripts/prepare/esm-bundle.ts
#	scripts/prepare/tsc.ts
#	scripts/release/__tests__/version.test.ts
#	scripts/tsconfig.json
#	scripts/yarn.lock

IanVS committed:
Update snapshot

IanVS committed:
Update globals

yannbf committed:
fix more tests

IanVS committed:
Merge remote-tracking branch 'origin/next' into norbert/vitest-for-monorepo

# Conflicts:
#	code/lib/cli/src/automigrate/fixes/__snapshots__/angular-builders.test.ts.snap
#	code/lib/cli/src/automigrate/fixes/add-react.test.ts
#	code/lib/cli/src/automigrate/fixes/autodocs-true.test.ts
#	code/lib/cli/src/automigrate/fixes/builder-vite.test.ts
#	code/lib/cli/src/automigrate/fixes/cra5.test.ts
#	code/lib/cli/src/automigrate/helpers/checkWebpack5Builder.test.ts
#	code/lib/cli/tsconfig.json
#	code/lib/manager-api/src/tests/refs.test.ts
#	code/lib/postinstall/package.json
#	code/lib/preview-api/src/modules/client-api/ClientApi.test.ts
#	code/lib/preview-api/src/modules/core-client/start.test.ts
#	code/lib/source-loader/package.json
#	code/package.json
#	code/presets/server-webpack/package.json
#	code/renderers/react/src/public-api.test.ts
#	code/renderers/vue/package.json
#	code/renderers/vue3/template/cli/ts-4-9/Page.stories.ts
#	code/yarn.lock
#	scripts/yarn.lock

IanVS committed:
Skip tests that are failing due to emotion theme

IanVS committed:
Convert new tests from jest to vitest

IanVS committed:
Fix eslint test

IanVS committed:
Update exports

IanVS committed:
Fix lockfile

IanVS commented:
I believe this is now ready for review.  

<img width="366" alt="image" src="https://github.com/storybookjs/storybook/assets/4616705/c6f831c4-72f5-4dbd-897b-7c77de3141e4">

I left a few comments above for some of the trickier or potentially problematic changes I made, which may now be collapsed.  So whoever reviews this, please give those a read.

IanVS commented:
Ergh, looks like there are still some CI failures.  But we're 🤏 close.

IanVS committed:
Fix lint errors

IanVS committed:
Update vitest to 1.0.1

IanVS committed:
Fix scripts tests

Snapshot indentation issues until https://github.com/vitest-dev/vitest/pull/4663 is released

IanVS committed:
Update snapshots

IanVS committed:
Unpin vite

IanVS committed:
Fix type errors

IanVS committed:
Merge branch 'next' into norbert/vitest-for-monorepo

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into norbert/vitest-for-monorepo

JReinhold committed:
re-add interopRequireDefault, use mock-require in unit tests

JReinhold committed:
revert wrong export statements

JReinhold committed:
increase test timeouts, add @types/mock-require

JReinhold committed:
increase timeout

JReinhold reviewed (changes requested):
I went through the whole thing (except for the gazillion jest -> vitest config changes TBH) and it looks great.

I only have a blocker on the snapshots that don't look right, I'll investigate it.

kasperpeulen committed:
Fix changelog test

ndelangen committed:
Merge branch 'next' into norbert/vitest-for-monorepo

valentinpalkovic committed:
Merge branch 'next' into norbert/vitest-for-monorepo

valentinpalkovic committed:
Fix Angular tests

valentinpalkovic committed:
Fix eslint

valentinpalkovic committed:
Fix Svelte tests

IanVS commented on code/renderers/svelte/src/__test__/Button.svelte:
```diff
@@ -1,8 +1,13 @@
 <script lang="ts">
-  import { createEventDispatcher } from 'svelte';
+  const { createEventDispatcher } = require('svelte');
```
This seems strange, I thought svelte was all-in on ESM.  Why was this change needed?

IanVS replied:
This is what I did previously to get those tests passing: https://github.com/storybookjs/storybook/pull/24415/commits/ca0394d77c0d41009181e73848634cdd6a3bd3c8

I see now there's an `// eslint-disable-next-line import/no-unresolved` comment above the import, so maybe that has something to do with it?  Why wouldn't that import be resolved?  And if it's not, it would explain why the files can't be parsed.

yannbf replied:
Seems like this was not needed after all, it was causing issues which got solved after undoing these changes

valentinpalkovic committed:
Fix test for Windows

yannbf committed:
fix types in compodoc tests

yannbf committed:
remove unnecessary directories

yannbf committed:
remove jest leftovers

yannbf committed:
temporarily debug windows test

yannbf committed:
undo function calls

yannbf committed:
temporarily skip type checks on public-types.test.ts

yannbf committed:
Merge branch 'next' into norbert/vitest-for-monorepo

yannbf committed:
fix helper tests on windows

yannbf committed:
migrate tests from jest to vitest

yannbf committed:
remove unnecessary ts ignore comment

yannbf committed:
increase test timeout

yannbf committed:
temporarily skip some ci checks

yannbf committed:
undo debugging checks

yannbf committed:
update misc code

yannbf committed:
undo svelte component changes

yannbf committed:
fix types in normalize-stories

yannbf commented on code/lib/core-common/src/utils/__tests__/normalize-stories.test.ts:
```diff
@@ -17,7 +18,7 @@ expect.addSnapshotSerializer({
 
 expect.extend({
-  toMatchPaths(regex: RegExp, paths: string[]) {
+  toEqual(regex: RegExp, paths: string[]) {
```
fixed!

IanVS commented on code/lib/core-common/vitest.d.ts:
```diff
@@ -0,0 +1,11 @@
+import 'vitest';
+
+interface CustomMatchers<R = unknown> {
+  toMatchPaths(paths: string[]): R;
```
The `expect.extend` should probably move to a test-setup file so it's available for every test, if we're adding the types like this.  It would be nice if there was a way to handle the types for an extends in just one test, though.

yannbf replied:
sounds good to me!

yannbf commented on code/renderers/react/src/docs/jsxDecorator.test.tsx:
```diff
@@ -120,10 +122,10 @@ describe('renderJsx', () => {
-        I am forwardRef!
-      </MyExoticComponent>
-    `);
+        <MyExoticComponent2>
+          I am forwardRef!
+        </MyExoticComponent2>
```
it's because the component is called MyExoticComponent and the function which is forwarded is also called MyExoticComponent

yannbf committed:
final cleanup

yannbf committed:
update contribution guide

yannbf committed:
fix typo
